solutions/usaco-894.mdx

```diff
@@ -19,222 +19,266 @@ It can be shown that the tree can also be colored accordingly in
+		min_types = max(min_types, t);
 	}
-	cout << maxAns << "\n";
+	
+	cout << min_types << endl;
 }
 ```
 
 </CPPSection>
-
 <JavaSection>
 
 ```java
 import java.io.*;
 import java.util.*;
-public class Main {
-	public static int[] ans;
-	public static List<Integer>[] adj;
+
+public class Planting {
+	static int[] neededTypes;
+	static List<Integer>[] neighbors;
+
 	public static void main(String[] args) throws IOException {
-		BufferedReader f = new BufferedReader(new FileReader("planting.in"));
-		PrintWriter pw = new PrintWriter(new BufferedWriter(new FileWriter("planting.out")));
-		StringTokenizer st = new StringTokenizer(f.readLine());
-		int farms = Integer.parseInt(st.nextToken());
-		adj = new List[farms];
-		for (int i = 0; i < farms; i++) {
-			adj[i] = new ArrayList<>();
+		Kattio io = new Kattio("planting");
+		
+		int fieldNum = io.nextInt();
+
+		neighbors = new List[fieldNum];
+		for (int f = 0; f < fieldNum; f++) {
+			neighbors[f] = new ArrayList<>();
 		}
-		for (int i = 0; i < farms - 1; i++) {
-			st = new StringTokenizer(f.readLine());
-			int a = Integer.parseInt(st.nextToken());
-			int b = Integer.parseInt(st.nextToken());
-			adj[a - 1].add(b - 1);
-			adj[b - 1].add(a - 1);
+		for (int i = 0; i < fieldNum - 1; i++) {
+			int field1 = io.nextInt();
+			int field2 = io.nextInt();
+			neighbors[--field1].add(--field2);
+			neighbors[field2].add(field1);
 		}
-		ans = new int[farms];
-		ans[0] = 1;
-		dfs(0, 0);
-		int max = 0;
-		for (int i : ans) {
-			max = Math.max(max, i);
+
+		neededTypes = new int[fieldNum];
+		neededTypes[0] = 1;
+
+		processFields(0, 0);
+		
+		int minTypes = 0;
+		for (int t : neededTypes) {
+			minTypes = Math.max(minTypes, t);
 		}
-		pw.println(max);
-		pw.close();
+
+		io.println(minTypes);
+		io.close();
 	}
-	public static void dfs(int self, int parent) {
-		int len = adj[self].size();
-		int g = 1;
-		for (int i : adj[self]) {
-			if (i != parent) {
-				while (g == ans[self] || g == ans[parent]) g++;
-				ans[i] = g;
-				dfs(i, self);
-				g++;
+
+	static void processFields(int at, int prev) {
+		int typeNum = 1;
+		for (int n : neighbors[at]) {
+			if (n == prev) {
+				continue;
+			}
+			while (typeNum == neededTypes[at] || typeNum == neededTypes[prev]) {
+				typeNum++;
 			}
+			neededTypes[n] = typeNum;
+			processFields(n, at);
+			typeNum++;
 		}
 	}
+
+	//CodeSnip{Kattio}
 }
 ```
 
 </JavaSection>
+<PySection>
+
+Implementing recursion as normal results in stack overflow for the last test case
+($10^5$ fields connected together in a single chain). To remedy this, we use our own
+stack.
+
+```py
+with open('planting.in') as read:
+	field_num = int(read.readline())
+	neighbors = [[] for _ in range(field_num)]
+	for _ in range(field_num - 1):
+		field1, field2 = [int(i) - 1 for i in read.readline().split()]
+		neighbors[field1].append(field2)
+		neighbors[field2].append(field1)
+
+needed_types = [0 for _ in range(field_num)]
+needed_types[0] = 1
+
+stack = [(0, 0)]
```

Comment:
i think `todo` might be a better name (completely up to you)

Reply:
makes sense, will do